bookkeeper-server/src/main/java/org/apache/bookkeeper/util/DoubleByteBuf.java

```diff
@@ -361,7 +362,7 @@ public ByteBuffer internalNioBuffer(int index, int length) {
 
     @Override
     public ByteBuffer nioBuffer(int index, int length) {
-        ByteBuffer dst = ByteBuffer.allocate(length);
+        ByteBuffer dst = isDirect() ? ByteBuffer.allocateDirect(length) : ByteBuffer.allocate(length);
```

Comment:
this is the only really necessary change, the other changes are just optimizations